test/refactoring/ConvertAsync/convert_to_continuation.swift

```diff
@@ -341,6 +448,100 @@ func testResultFromValueAndError(completionHandler: (Result<Int, Error>) -> Void
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:       guard let theResult1 = theResult else {
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:         fatalError("Expected non-nil result 'theResult1' for nil error")
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:       }
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:       continuation.resume(returning: (theResult1, "hi"))
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:     }
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:   }
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT: }
+
+// RUN: %refactor-check-compiles -convert-to-async -dump-text -source-filename %s -pos=%(line+1):1 | %FileCheck -check-prefix=USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL %s
+func testUseOptionalResultValueAfterCompletionHandlerCall(completion: (Int?, String?, Error?) -> Void) {
+  withoutAsyncAlternativeThrowing { (theResult, error) in
+    completion(theResult, "hi", nil)
+    print(theResult.map { $0 + 1 } as Any)
+  }
+}
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL:      func testUseOptionalResultValueAfterCompletionHandlerCall() async throws -> (Int, String) {
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:   return try await withCheckedThrowingContinuation { continuation in 
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:     withoutAsyncAlternativeThrowing { (theResult, error) in
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:       guard let theResult1 = theResult else {
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:         fatalError("Expected non-nil result 'theResult1' for nil error")
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:       }
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:       continuation.resume(returning: (theResult1, "hi"))
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:       print(theResult.map { $0 + 1 } as Any)
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:     }
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:   }
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT: }
+
+// We shouldn't need to unwrap `theResult` twice here, but the example is silly and I don't care too much.
+// RUN: %refactor-check-compiles -convert-to-async -dump-text -source-filename %s -pos=%(line+1):1 | %FileCheck -check-prefix=PASS-SAME-RESULT-TWICE %s
+func testPassSameResultTwice(completion: (Int?, Int?, Error?) -> Void) {
+  withoutAsyncAlternativeThrowing { (theResult, error) in
+    completion(theResult, theResult, nil)
+  }
+}
+// PASS-SAME-RESULT-TWICE:      func testPassSameResultTwice() async throws -> (Int, Int) {
+// PASS-SAME-RESULT-TWICE-NEXT:   return try await withCheckedThrowingContinuation { continuation in 
+// PASS-SAME-RESULT-TWICE-NEXT:     withoutAsyncAlternativeThrowing { (theResult, error) in
+// PASS-SAME-RESULT-TWICE-NEXT:       guard let theResult1 = theResult else {
```

Comment:
How hard would this be to avoid? If not much effort it'd be good to fix, but probably an uncommon case so I'm fine with leaving.

Reply:
Probably not too much if we only worry about the *exact same* argument being passed to the completion handler twice. But I think there’s very little benefit in it (who passes the same argument to the completion handler twice but me?) and I’m more worried that the logic will introduce bugs in the common cases.